### ISSUE TITLE
In Windows: Fips provider failed to load, Could not find the shared library 

### ISSUE THREAD

**Issue by DStarts23** [openssl/openssl#25350]
We need to migrate from Openssl 1.0.2.u version to Openssl 3.0.8 (Fips 140-2 validated version). For this purpose, I had compiled the Openssl 3.0.8 version on Jenkins using Windows Batch commands tool to create openssl fips enable:

```
perl Configure VC-WIN32 --prefix="%WORKSPACE%\x86" enable-fips

@echo Generating FIPS capable OpenSSL DLLs
nmake

@echo Running self test
nmake test

@echo Copying files to the delivery folder (prefix)
nmake install
```

It generates the following files:
```
bin --> libcrypto-3.dll, libcrypto-3.pdb, libssl-3.dll, libssl-3.pdb, openssl.exe, openssl.pdb
include --> openssl -> header files
lib --> engines3 -> capi.dll, capi.pdb, loader_attic.dll, loader_attic.pdb, padlock.dll, padlock.pdb
   |-->ossl-modules -> fips.dll, fips.pdb, legacy.dll, legacy.pdb
   |--> libcrypto.lib, libssl.lib
```

I created a console application, and add those Header files in External Dependencies  using **Project Properties -> C/C++ -> General ->Additional Include directories -> (Include folder)** 
Added libraries using **Project Properties -> Linker -> Libraries -> Lib (folder)**
Added Additional Libraries: **Properties -> Linker -> Additional Libraries -> libcrypto.lib, libssl.lib**
Also added the fips.dll, openssl.cnf file, fipsmodule.cnf, libcrypto-3.dll, libssl-3.dll with my project Exe in bin\x86\Debug folder.

And write down the following code:

``` C++
 OSSL_PROVIDER* fips_provider = nullptr;
 OSSL_PROVIDER* default_provider = nullptr;
 
 // Load the default provider first
 default_provider = OSSL_PROVIDER_load(libctx, "default");
 if (default_provider == nullptr) {
     std::cerr << "Failed to load default provider" << std::endl;
     ERR_print_errors_fp(stderr);
 }

 // Load the FIPS provider
 fips_provider = OSSL_PROVIDER_load(NULL, "fips");
 if (fips_provider == nullptr) {
     std::cerr << "Failed to load FIPS provider" << std::endl;
    ERR_print_errors_fp(stderr);
 }
```

This code somehow load the default and base provider, but unable to load the Fips. I am getting these errors:

```
**error:12800067:DSO support routines::could not load the shared library**
Failed to load FIPS provider OPENSSL_Uplink(6BF7F380,08): no OPENSSL_Applink
```

Then add Applink.c file to resolve this issue, and build again, But get this error:

```
AC260000:error:1C8000D5:Provider routines:SELF_TEST_post:missing config data:providers\fips\self_test.c:290: AC260000:error:1C8000E0:Provider routines:ossl_set_error_state:fips module entering error state:providers\fips\self_test.c:388: AC260000:error:1C8000D8:Provider routines:OSSL_provider_init_int:self test post failure:providers\fips\fipsprov.c:707: AC260000:error:078C0105:common libcrypto routines:provider_init:init fail:crypto\provider_core.c:932:name=fips
```

My openssl.cnf file configuration is written as:

```
# For FIPS
# Optionally include a file that is generated by the OpenSSL fipsinstall
# application. This file contains configuration data required by the OpenSSL
# fips provider. It contains a named section e.g. [fips_sect] which is
# referenced from the [provider_sect] below.
# Refer to the OpenSSL security policy for more information.
# .include fipsmodule.cnf

[openssl_init]
providers = provider_sect

# List of providers to load
[provider_sect]
default = default_sect
# The fips section name should match the section name inside the
# included fipsmodule.cnf.
fips = fips_sect

# If no providers are activated explicitly, the default one is activated implicitly.
# See man 7 OSSL_PROVIDER-default for more details.
#
# If you add a section explicitly activating any other provider(s), you most
# probably need to explicitly activate the default provider, otherwise it
# becomes unavailable in openssl.  As a consequence applications depending on
# OpenSSL may not work correctly which could lead to significant system
# problems including inability to remotely access the system.
[default_sect]
# activate = 1
```

The content of fipsmodule.cnf is:

```
[fips_sect]
activate = 1
conditional-errors = 1
security-checks = 1
module-mac = 6C:B9:1C:63:66:96:EE:92:45:92:A5:C2:A3:AD:57:36:DB:40:DB:3E:F6:CC:33:AF:0E:B6:28:83:E8:0D:A8:1E
```

Plus where I need to place the openssl.cnf, fips.dll and libcrypto-3.dll, libssl-3.dll files in my project Directory? I tried the other ways as well which have been discussed in https://docs.openssl.org/3.0/man7/fips_module/ but it always show NULL 0x0000000<NULL> in fipsProvider variable, and goes through some other issues that has been discussed here in GitHub, but didn't work on myside. 
Can you please help me in finding out how ithout installing Openss 3 on my machine, I can load these libraries in Windows?

**Comment by levitte** [comment]
You must uncomment this line in openssl.cnf:

```
# .include fipsmodule.cnf
```

**Comment by levitte** [comment]
Ah, right, it might work better if you define an environment variable `OPENSSL_MODULES` with the path to that directory.  Otherwise, it tries to load them from a hardcoded path (you can find out what it is with `openssl info -modulesdir`)

**Comment by AimaMaqsood108** [comment]
@levitte The issue is this command  `openssl info -modulesdir` is referencing to the Jenkins Environment workspace as:

`D:\Jenkins\Workspace\Our_Project\Fips_Build_Job\Delivery\x86\lib\ossl-modules`

And the ossl-modules contain these files: fips.dll, fips.pdb, legacy.dll, legacy.pdb 
What I do now, copy the content of ossl-modules folder and paste in separate directory, then create the Environment System variable as:
![image](https://github.com/user-attachments/assets/1e0d5ffe-16c7-426b-a777-fac969a24470)

But still it is showing NULL. I have not installed Openssl 3 on my machine, instead compiling it using Jenkins, and getting those files to use them in my program.

**Comment by levitte** [comment]
Okie...  looking back at the error, it looks like it does actually load, but selftest doesn't see the config data.  There are two more envvars that come into play:

- **OPENSSL_CONF** - the path (including file name) of openssl.cnf
- **OPENSSL_CONF_INCLUDE** - the directory for config includes (i.e. the `.include` you uncommented)

**Comment by AimaMaqsood108** [comment]
I created these env variables as:
![image](https://github.com/user-attachments/assets/e9134a70-0868-4683-8e73-c25954790784)

And run the program but still getting error:
![image](https://github.com/user-attachments/assets/9437bbe8-d46b-4650-9b24-369ff99a13d3)

The content in fipsmodule.cnf is:
![image](https://github.com/user-attachments/assets/da33f84e-6f8b-4c08-9826-39c903095162)

And the content in openssl.cnf file is:
# For FIPS
# Optionally include a file that is generated by the OpenSSL fipsinstall
# application. This file contains configuration data required by the OpenSSL
# fips provider. It contains a named section e.g. [fips_sect] which is
# referenced from the [provider_sect] below.
# Refer to the OpenSSL security policy for more information.
.include fipsmodule.cnf

[openssl_init]
providers = provider_sect

# List of providers to load
[provider_sect]
default = default_sect
fips = fips_sect

# The fips section name should match the section name inside the
# included fipsmodule.cnf.

# If no providers are activated explicitly, the default one is activated implicitly.
# See man 7 OSSL_PROVIDER-default for more details.
#
# If you add a section explicitly activating any other provider(s), you most
# probably need to explicitly activate the default provider, otherwise it
# becomes unavailable in openssl.  As a consequence applications depending on
# OpenSSL may not work correctly which could lead to significant system
# problems including inability to remotely access the system.
[default_sect]
# activate = 1

**Comment by nhorman** [comment]
From the errors it appears your config file still isn't getting completely loaded, as the error you're seeing is encountered when the module_mac value in the fips provider isn't getting parsed at startup.

Looking at your config file, I would guess that your missing this line:
```
 openssl_conf = openssl_init
```
In the global namespace of the file.  Without that, the config parser won't know that [openssl_init] is the entry point to your parse tree.  

see https://docs.openssl.org/3.0/man5/config/#openssl-library-configuration for details

**Comment by AimaMaqsood108** [comment]
@nhorman This line has been already the part of openssl.cnf file as:
`#
# OpenSSL example configuration file.
# See doc/man5/config.pod for more info.
#
# This is mostly being used for generation of certificate requests,
# but may be used for auto loading of providers

# Note that you can include other files from the main configuration
# file using the .include directive.
#.include filename

# This definition stops the following lines choking if HOME isn't
# defined.
HOME			= .

 # Use this in order to automatically load providers.
openssl_conf = openssl_init

# Comment out the next line to ignore configuration errors
config_diagnostics = 1`

**Comment by nhorman** [comment]
Your comments [here](https://github.com/openssl/openssl/issues/25350#issuecomment-2324420406) and [here](https://github.com/openssl/openssl/issues/25350#issue-2500025110) seem to contradict that. 

I suggest that you include the full config file here please so we can see what it looks like currently, and add the conf_diagnostics statement to it, so that we can follow its parsing during your application run.

**Comment by AimaMaqsood108** [comment]
@nhorman Okay let me share the both config files, as the first commented openssl.cnf & fipsmodule.cnf files are not modified.

Fipsmodule.cnf:
[fipsmodule.json](https://github.com/user-attachments/files/16837093/fipsmodule.json)

openssl.cnf file content:
[openssl.json](https://github.com/user-attachments/files/16837103/openssl.json)

**Comment by nhorman** [comment]
I'm confused.  The openssl.cnf file that you attached appears to be the example config file provided in the source, which looks _very_ different from the one you provided in your comments above.  Which one are you using?

**Comment by AimaMaqsood108** [comment]
The current one which I have shared with you openssl.cnf (Json) format
file. After having discussion with Richard, I updated the comment one file
and shred the latest (json one) with you.

**Comment by nhorman** [comment]
ok, then I would think the next step would be to confirm that we're using the config file we think we are.  Can you set a break point in CONF_modules_load_file_ex and re-run your application, interrogating the value of the filename parameter passed there?  Further, can you confirm that the file which you provided above matches the contents of the file name in that path?

**Comment by AimaMaqsood108** [comment]
I checked the provider.h file from my Project External Dependencies, and it doesn't contain this API: CONF_modules_load_file_ex
Plus the build output shows the Fips.dll loaded and then unloaded and exit the application.
![image](https://github.com/user-attachments/assets/2ec4408d-00c9-4b1d-b631-2e481df9f42b)

**Comment by nhorman** [comment]
The api is clearly declared in conf.h and is available in the libcrypto.so.3 library.  If you've build with debugging enabled, you should be able to load it

Though the output you've provided above gives us some more clues as to whats going on:
```
TestFIPSProvider.exe (Win32): Loaded 'C:Projects/ir-arc-crypto\CryptOn\Win32\Debug\fips.dll
```
That path in no way matches where you installed your copy of openssl to in your initial comment, nor does it match the path of OPENSSL_MODULES that you set above.  It seems like you are using a mismatched environment, potentially loading a fips provider that doesn't match the integrity check of your fipsmodule.cnf

**Comment by AimaMaqsood108** [comment]
@nhorman  Okay, as you said in the last comment, I added the conf.h file and run the application to get the value for `CONF_modules_load_file_ex` as:

` OSSL_LIB_CTX* libctx = nullptr; 
const char* filename = "C:\\Program Files (x86)\\OpenSSL\\config\\openssl.cnf";
 const char* appname = "TestFIPSProvider";
 unsigned long flags = 0; // default behavior
 int config_Value = CONF_modules_load_file_ex(libctx, filename,appname, flags);`
 
 And this API returns, config_value=1, but didn't load the providers.
 For the path thing, as you have mentioned, let me clear everything from start what changes I have done in my application:
 
 Created System Variables: 
 
![image](https://github.com/user-attachments/assets/12ae878a-18c4-4e99-b48e-ab5208638b7d)
config folder have two files (which I had already shared with you in json format): openssl.cnf, fipsmodule.cnf
ossl-modules folder have: fips.dll, fips.pdb, legacy.dll, legacy.pdb

Added Dll's in Test Application:
I have added the .cnf and dll files also in test applications under x86\Debug folder:
![image](https://github.com/user-attachments/assets/c10a98e9-26bf-4e53-ad70-189964e79bb1)

Now can you tell me which path I am setting wrong here?

**Comment by nhorman** [comment]
Unfortunately, no, or at least not entirely

What I can tell you
1) Your installation is properly loading your config file from C:\Program Files (x86)\OpenSSL\config\openssl.cnf, which matches what your OPENSSL_CONF environment variable says, so thats good
2) You indicate that you added a cnf file in x86\debug folder.  I can't tell, since thats a truncated path, but doing this seems moot, as the cnf file isn't getting loaded from any directory with x86\debug in it
3) You're loading fips.dll from C:\Projects/ir-arc-crypto\CryptOn\Win32\Debug\fips.dll based on your prior comment, which _doesn't_ match OPENSSL_MODULES

So, it would seem to me that the config file in C:\Program Files (x86)\OpenSSL\config\openssl.cnf, and somehow that file is redirecting your module loads to come from C:\Projects/ir-arc-crypto\CryptOn\Win32\Debug\

That would imply that you have a path directive specifying the location of your fips module.

I have to ask again, you've provided several openssl.cnf files.  Which one matches the content in the file C:\Program Files (x86)\OpenSSL\config\openssl.cnf.  Can you post the content of that file specifically?

**Comment by AimaMaqsood108** [comment]
This one is the final openssl.cnf  content of the file I am using: https://github.com/openssl/openssl/issues/25350#issuecomment-2324578030

**Comment by nhorman** [comment]
well, it depends what you want to load.  If you want to put the provider dlls there, you can, you just have to make sure OSSL_MODULES match that location. 

Though that doesn't seem to be whats going on here.  Based on your debugger output [here](https://github.com/openssl/openssl/issues/25350#issuecomment-2324901668), you're loading the provider from C:Projects/ir-arc-crypto\CryptOn\Win32\Debug\fips.dll

that can happen for a number of reasons, but the usual suspects seem to be covered here.  

the bottom line is that you're loading a fips dll that (ostensibly) doesn't match the mac values in your generated config file, so its incumbent on us to figure out why you're loading from that path, rather than the one specified by OPENSSL_CONF_INCLUDE

Can you set a breakpoint in module_load_dso and see what the path value is that gets passed to the call to DSO_load in there?

**Comment by levitte** [comment]
``` C
const char* appname = "TestFIPSProvider";
```

This will use `TestFIPSProvider` instead of `openssl_conf` as a starting point for the configuration.  This must be reflected in the configuration file you use.  Alternatively, NULL for appname to `CONF_modules_load_file_ex()`, and it will use the default name `openssl_conf`, which is what you have in the config file you shared with us.

**Comment by AimaMaqsood108** [comment]
@nhorman  You are write about the path issue, it was my mistake, the dependencies I have added in my Test Project refers to some old libraries that are in this path: C:Projects/ir-arc-crypto\CryptOn\Win32\Debug\fips.dll

After removing those previous paths, I started getting this error:

`400F0000:error:80000002:system library:BIO_new_file:No such file or directory:crypto\bio\bss_file.c:67:calling fopen(openssl.cnf, rb)
400F0000:error:10000080:BIO routines:BIO_new_file:no such file:crypto\bio\bss_file.c:75:
400F0000:error:07000072:configuration file routines:def_load:no such file:crypto\conf\conf_def.c:179:
400F0000:error:80000002:system library:process_include:No such file or directory:crypto\conf\conf_def.c:805:calling stat(fipsmodule.cnf)
400F0000:error:07800069:common libcrypto routines:provider_conf_load:provider section error:crypto\provider_conf.c:156:section=fips_sect not found
400F0000:error:0700006D:configuration file routines:module_run:module initialization error:crypto\conf\conf_mod.c:270:module=providers, value=provider_sect retcode=-1
400F0000:error:12800067:DSO support routines:win32_load:could not load the shared library:crypto\dso\dso_win32.c:108:filename(C:\Projects\ir-arc-crypto\CryptOn\Win32\Debug\fips.dll)
400F0000:error:12800067:DSO support routines:DSO_load:could not load the shared library:crypto\dso\dso_lib.c:152:
400F0000:error:07880025:common libcrypto routines:provider_init:reason(524325):crypto\provider_core.c:912:name=fips`

But it is still pointing to the older project. Can you please tell me the steps to add the Opnessl compiled files into the projects correctly? It's very confusing for me now as well :(

**Comment by nhorman** [comment]
thats something I'm afraid you will have to research, we don't typically use visual studio here.

All I can tell you from the above though is that it appears that your definitions for the OPENSSL_CONF and other environment variables aren't getting picked up properly (otherwise those paths would have appeared in those error messages)

**Comment by AimaMaqsood108** [comment]
Okay, I'll do the research and will try to resolve this path issue as well, but can you tell me the  openssl.cnf file I have shared with you above in this comment https://github.com/openssl/openssl/issues/25350#issuecomment-2324578030 is okay from the point of enabling Fips? Like my requirement is by default, default provider will do encryption/decryption, and if user enables the Fips mode then Fips provider will be used.

**Comment by nhorman** [comment]
as long as the the fipsmodule.cnf file contains the correct mac values (as produced by the fipsinstall command) is correct, yes, those files should generally speaking, be fine.  We will know for certain once you get them loading properly.  We can tackle any subsequent issues with the config once we get there

**Comment by DStarts23** [comment]
Hello, @nhorman , finally I am able to load the FIPS Provider in Visual Studio on Windows, that can automatically load the Default provider and then on user request, loads the FIPS Provider. The issue was about Environment variables, which needs to be explicitly initialized in the Visual Studio under "project properties" -> "Debugging" \ "Release" Mode. Then in openssl.cnf file, need to do these changes:

`.include "path/to/fipsmodule.cnf"

[openssl_init]
providers = provider_sect

# List of providers to load
[provider_sect]
default = default_sect
# The fips section name should match the section name inside the
# included fipsmodule.cnf.
# fips = fips_sect

# If no providers are activated explicitly, the default one is activated implicitly.
# See man 7 OSSL_PROVIDER-default for more details.
#
# If you add a section explicitly activating any other provider(s), you most
# probably need to explicitly activate the default provider, otherwise it
# becomes unavailable in openssl.  As a consequence applications depending on
# OpenSSL may not work correctly which could lead to significant system
# problems including inability to remotely access the system.
[default_sect]
activate = 1

[fips_sect]
module = path/to/fips-dll
`
The last thing I want to confirm is these digest algorithms:
```
OpenSSL_add_all_algorithms();
OpenSSL_add_all_digests();
OpenSSL_add_all_ciphers();
```
In openssl 1.0.2.x version, we were loading all the libraries, do these things need to be loaded in OpenSSL 3.0.8 version as well?

**Comment by nhorman** [comment]
I don't believe those need to be called in 3.x, as they can be individually fetched via the EVP_[CIPHER|MD]_fetch calls.

**Comment by DStarts23** [comment]
Okay , Thank you for your help.
